### PR TITLE
Check for operator version and basic image update logic were implemented

### LINKED ISSUE #638: Add check for operator versions

### PR DESCRIPTION
**Pull request by DoodgeMatvey**
Closes #638

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -121,6 +123,17 @@ func main() {
 
 	s := scheduler.NewScheduler(log.Log.WithValues("component", "scheduler"))
 
+	// TODO: take this variable from helm env
+	upgradeCheckingDisabled := true
+	if !upgradeCheckingDisabled {
+		setupLog.Info("upgrade checking enabled")
+
+		err = upgradecheck.StartUpgradeCheckJob(context.TODO(), mgr.GetClient(), s)
```
Comment (ribaraka) [inline main.go:131]: 1. upgradecheck.StartUpgradeCheckJob -> upgradecheck.Start()
2. do you actually need a context?

File: pkg/models/upgrade_check.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package models
+
+import "time"
+
+const (
+	UpgradeCheckInterval   = time.Second * 120
```
Comment (ribaraka) [inline pkg/models/upgrade_check.go:6]: add this to scheduler

File: pkg/models/upgrade_check.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+	OperatorUpgradeChecker = "operatorUpgradeChecker"
+)
+
+type DockerTagsResponse struct {
```
Comment (ribaraka) [inline pkg/models/upgrade_check.go:10]: move it to `upgradechace` pkg

File: pkg/upgradecheck/upgrade_check.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/instaclustr/operator/pkg/scheduler"
+)
+
+func StartUpgradeCheckJob(ctx context.Context, client client.Client, s scheduler.Interface) error {
+	job := newUpgradeCheckJob(ctx, client)
+	return s.ScheduleJob(models.OperatorUpgradeChecker, models.UpgradeCheckInterval, job)
+}
+
+func newUpgradeCheckJob(ctx context.Context, client client.Client) scheduler.Job {
+	l := log.FromContext(ctx, "components", "WatchStatusJob")
```
Comment (ribaraka) [inline pkg/upgradecheck/upgrade_check.go:28]: WatchStatusJob -> OperatorAutoUpdateJob

File: pkg/upgradecheck/upgrade_check.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+	return func() error {
+		// TODO: change from dockerhub to custom endpoint
+		latestTag, err := GetLatestDockerImageTag("icoperator/instaclustr-operator")
+		if err != nil {
+			l.Error(err, "unable to get latest docker image tag")
+		}
```
Comment (ribaraka) [inline pkg/upgradecheck/upgrade_check.go:35]: you forgot `return error `

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: pkg/upgradecheck/upgrade_check.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		err = UpdateImageTagIfNeeded(ctx, l, client, latestTag)
+		if err != nil {
+			l.Error(err, "unable to update current image tag")
+		}
+
+		return nil
+	}
+}
+
+func GetLatestDockerImageTag(imageName string) (string, error) {
```
Comment (testisnullus) [inline pkg/upgradecheck/upgrade_check.go:46]: Let's make functions that are not used in other packages unexportable

File: pkg/upgradecheck/upgrade_check.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	url := fmt.Sprintf("https://registry.hub.docker.com/v2/repositories/%s/tags?page_size=1", imageName)
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (testisnullus) [inline pkg/upgradecheck/upgrade_check.go:55]: Let's use `io.ReadAll()` here and in other cases, cause the `ioutil` package is deprecated

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: pkg/upgradecheck/upgrade_check.go (hunk 7 above)
Comment (DoodgeMatvey) [reply pkg/upgradecheck/upgrade_check.go]: Fixed

File: pkg/upgradecheck/upgrade_check.go (hunk 6 above)
Comment (DoodgeMatvey) [reply pkg/upgradecheck/upgrade_check.go]: Fixed

File: pkg/upgradecheck/upgrade_check.go (hunk 5 above)
Comment (DoodgeMatvey) [reply pkg/upgradecheck/upgrade_check.go:35]: Agree, fixed

File: pkg/upgradecheck/upgrade_check.go (hunk 4 above)
Comment (DoodgeMatvey) [reply pkg/upgradecheck/upgrade_check.go]: Renamed to `UpgradeCheckJob`. WDYT?

File: pkg/models/upgrade_check.go (hunk 3 above)
Comment (DoodgeMatvey) [reply pkg/models/upgrade_check.go:7]: I think it is better to keep structs in models pkg

File: pkg/models/upgrade_check.go (hunk 2 above)
Comment (DoodgeMatvey) [reply pkg/models/upgrade_check.go]: Replaced

File: main.go (hunk 1 above)
Comment (DoodgeMatvey) [reply main.go:131]: There are can be other jobs in this package, so I would like to keep it.
Context we need for logging and getting operator deployment

**Review by testisnullus — APPROVED** [review]

**Review by ribaraka — APPROVED** [review]

**Pull request merged**